src/powershell/Program.cs

```diff
@@ -19,7 +20,232 @@ public sealed class ManagedPSEntry
+        {
+            loginIndex = -1;
+
+            // Parameter comparison strings, stackalloc'd for performance
+            for (int i = 0; i < args.Length; i++)
+            {
+                string arg = args[i];
+
+                // Must look like '-<name>'
+                if (arg == null || arg.Length < 2 || arg[0] != '-')
+                { 
+                    continue;
+                }
+
+                // Check for "-Login" or some prefix thereof
+                if (IsParam(arg, "login", "LOGIN"))
+                {
+                    loginIndex = i;
+                    return true;
+                }
+
+                // After -File and -Command, all parameters are passed
+                // to the invoked file or command, so we can stop looking.
+                if (IsParam(arg, "file", "FILE")
+                    || IsParam(arg, "command", "COMMAND"))
+                {
+                    return false;
+                }
+            }
+
+            return false;
+        }
+
+        /// <summary>
+        /// Determines if a given parameter is the one we're looking for.
+        /// Assumes any prefix determines that parameter (true for -l, -c and -f).
+        /// </summary>
+        /// <param name="arg">The argument to check.</param>
+        /// <param name="paramToCheck">The lowercase name of the parameter to check.</param>
+        /// <param name="paramToCheckUpper">The uppercase name of the parameter to check.</param>
+        /// <returns></returns>
+        private static bool IsParam(
+            string arg,
+            string paramToCheck,
+            string paramToCheckUpper)
+        {
+            // Quick fail if the argument is longer than the parameter
+            if (arg.Length > paramToCheck.Length + 1)
+            {
+                return false;
+            }
+
+            // Check arg chars in order and allow prefixes
+            for (int i = 1; i < arg.Length; i++)
+            {
+                if (arg[i] != paramToCheck[i-1]
+                    && arg[i] != paramToCheckUpper[i-1])
+                {
+                    return false;
+                }
+            }
+
+            return true;
+        }
+
+        /// <summary>
+        /// Create the exec call to /bin/{ba}sh -l -c 'pwsh "$@"' and run it.
+        /// </summary>
+        /// <param name="args">The argument vector passed to pwsh.</param>
+        /// <param name="loginArgIndex">The index of -Login in the argument vector.</param>
+        /// <returns>
+        /// The exit code of exec if it fails.
+        /// If exec succeeds, this process is overwritten so we never actually return.
+        /// </returns>
+        private static int ExecPwshLogin(string[] args, int loginArgIndex)
+        {
+            // We need the path to the current pwsh executable
+            string pwshPath = System.Diagnostics.Process.GetCurrentProcess().MainModule.FileName;
```

Comment:
We might be able to find a faster API to get the path to pwsh.exe or `$PSHOME` here, but need advice on what that is. One I was looking at was `Assembly.GetExecutingAssembly().Location`

Reply:
I've tried this now, and it's slower. Keeping the `GetCurrentProcess` version

Reply:
We could compare with `Environment.CommandLine`.

Reply:
That one might be more work; it depends on the invocation and may start with a relative path and contain arguments. In my current session `[Environment]::CommandLine`:

```
"C:\Program Files\PowerShell\7-preview\pwsh.dll" -WorkingDirectory ~
```

Also, like you can see in the code in this PR for `exec`, you can set `$0`, so that could almost be anything.

I also thought about `[Environment]::CurrentDirectory` and `[System.IO.Directory]::GetCurrentDirectory` but again those can both be changed.

Reply:
If we say about login shell scenario I guess we have absolute path.

Reply:
Since you have to inspect `/proc` and call `sysctl()` anyways, you should use that to get this info

Reply:
Yes I had the same thought! Trying to work out if that's possible on macOS